### DISCUSSION TITLE
What's Taichi's take on NVIDIA's Warp?

### DISCUSSION THREAD

**Discussion by Ruibin-Liu** [taichi-dev/taichi#8184] [General]
[NVIDIA's Warp](https://github.com/NVIDIA/warp) says:

> Compared to Taichi, Warp uses C++/CUDA as an intermediate representation, which makes it convenient to implement and expose low-level routines. In addition, we are building in data structures to support geometry processing (meshes, sparse volumes, point clouds, USD data) as first-class citizens that are not exposed in other runtimes.

I don't find it incorrect but does Taichi has plans to enhance in the two points mentioned? Maybe the first one is related to https://github.com/taichi-dev/taichi/discussions/5446#discussioncomment-3170057?

**Comment by bobcao3** [comment]
Overall the biggest distinction as of now is that Taichi operates at a slightly higher level. E.g. implict loop parallelization, high level spatial data structures, direct interops with torch, etc.

We are trying to implement support for lower level programming styles to accommodate such things as native intrinsics, but we do think of those as more advanced optimization techniques, and at the same time we strive for easier entry and usage for beginners or people not so used to CUDA's programming model
